Support the --reftest-screenshot command line argument in the reftest implementation used by all browsers

rmhasan opened the pull request:
I am writing a test for Chromium where I need screenshots from passing reftests. The reftest will run after test data is loaded into Chromium's user directory. After the reftest runs, I need to compare the screenshots generated during the test run to baselines created from the last known good run of the reftest where the test data was not loaded. I can use the --reftest-screenshot command line argument to make the wpt test harness add screenshots made during passing reftests to the test results.

Bug:crbug/1285152

Commit:
Add option to add all screenshots created during a reftest to the test result

I am writing a test for Chromium where I need screenshots from passing
reftests. The reftest will run after test data is loaded into Chromium's
user directory. After the reftest runs, I need to compare the screenshots
generated during the test run to baselines created from the last known
good run of the reftest where the test data was not loaded.

Bug:crbug/1285152

jgraham commented:
There's already a gecko-specific `--reftest-screenshot=[always|fail|unexpected]` option. That's used as an optimisation for the gecko case where we can avoid serialising the screenshot entirely. But I think it would make more sense to extend that option to all browsers and use it to also control when to output the screenshot (the the gecko optimisation would be "don't serialise a screenshot we won't output", which makes sense).

In any case we should avoid the use of the word `artifact` here, since that has a specific meaning in the taskcluster CI we're using.

I also wonder if you have more details on the use case here; I assume needing to compare to the older screenshots is a one-off thing and not an integral part of the test?

Commit:
Use --reftest-screenshot to control which screenshots are added to
the test results.

rmhasan commented:
@jgraham I updated the PR so that RefTestImplmentation uses the --reftest-screenshot command line argument to decide when to add screenshots to the test result. I also removed the use of the word artifact from this PR and all the commits I made. I added more details of my use case to the PR message and commits. The test I am writing involves loading test data into chromium and then running reftests. After running the reftests I want to compare the screenshots taken to baseline screenshots created from the last known good run of the reftest where the test data was not loaded.

jgraham commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -357,12 +358,13 @@ class RefTestExecutor(TestExecutor):
 
     def __init__(self, logger, browser, server_config, timeout_multiplier=1, screenshot_cache=None,
-                 debug_info=None, **kwargs):
+                 debug_info=None, reftest_screenshot=None, **kwargs):
```
Let's default to "unexpected" since in practice this can't actually have a `None` value (we always set it in wptcommandline.py, with the no-argument default being set depending on which other parameters were supplied).

rmhasan replied:
done

jgraham commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -504,6 +507,8 @@ def run_test(self, test):
 
         stack = list(((test, item[0]), item[1]) for item in reversed(test.references))
         page_idx = None
```
So we depend on this being set in the loop below and if it's ever `None` that's now an error in the code. So I think we should remove the declaration which sets it to `None`.

rmhasan replied:
done

jgraham commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -515,45 +520,63 @@ def run_test(self, test):
+
             if is_pass:
+                if self.reftest_screenshot == "always":
+                    log_data.extend(
```
So unfortunately this doesn't really work as written.

Some log formatters assume that `reftest_screenshots` is a list of exactly 3 items (e.g. https://searchfox.org/mozilla-central/source/testing/mozbase/mozlog/mozlog/formatters/tbplformatter.py#258) so we can't just add every screenshot to the log output without breaking things.

With the current API we can only really take the last screenshot we get.

rmhasan replied:
That is fine with me. I changed the function to only add the last two reftests run to the test result dictionary.

jgraham commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -515,45 +520,63 @@ def run_test(self, test):
                     stack.extend(list(((nodes[1], item[0]), item[1])
                                       for item in reversed(nodes[1].references)))
                 else:
+                    test_result = {"status": "PASS", "message": None}
```
Note that if the test is expected to `FAIL` then we should output screenshots if it actually get as `PASS` status and `reftest_screenshot` is set to `"unexpected"`.

rmhasan replied:
Done

jgraham commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -515,45 +520,63 @@ def run_test(self, test):
+        elif (not self.reftest_screenshot or
+              self.reftest_screenshot == "fail" or
+              (self.reftest_screenshot == "unexpected" and
+               test.expected() == 'PASS')):
```
We should write the comparison with `expected()` like `test.expected() != "FAIL"` rather than assuming that only an expected `PASS` could require a screenshot if the result is `FAIL`.

rmhasan replied:
Done

Commit:
Remove setting of the reftest_screenshot executor arg in browsers/firefox.py

Commit:
Set the default value for the reftest_screenshot parameter to 'unexpected'

Commit:
Only log the last two screenshots taken during a reftest

rmhasan commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -30,6 +30,7 @@ def executor_kwargs(test_type, test_environment, run_info_data, **kwargs):
 
     if test_type in ("reftest", "print-reftest"):
         executor_kwargs["screenshot_cache"] = test_environment.cache_manager.dict()
+        executor_kwargs["reftest_screenshot"] = kwargs["reftest_screenshot"]
```
done

Commit:
Return -1 as the page index if the number of the screenshots for each reftest page are not equal

Commit:
Set the reftest_screenshot parameter's default value to 'unexpected' for the WebDriverRefTestExecutor constructor

jgraham commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -521,39 +524,43 @@ def run_test(self, test):
                 else:
+                    test_result = {"status": "PASS", "message": None}
+                    if (self.reftest_screenshot == "always" or
+                            self.reftest_screenshot == "unexpected" and
```
Style in this code is to indent this to align with the line above, not to add an extra level of indent.

rmhasan replied:
Done

jgraham commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -521,39 +524,43 @@ def run_test(self, test):
+        test_result =  {"status": "FAIL",
+                        "message": "\n".join(self.message)}
+        if (self.reftest_screenshot in ("always", "fail") or
+                self.reftest_screenshot == "unexpected" and
```
Same indent comment

rmhasan replied:
Done

jgraham reviewed (commented):
Just a couple of minor style issues.

Commit:
Use the same line indent size in the multiline if statement expression

jgraham commented:
```
./wptrunner/wptrunner/executors/base.py:557:22: E222 multiple spaces after operator
```

Commit:
Remove extra space around the = operator in base.py on line 557
